File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/web/CurrentRequestContext.java
package com.github.arhor.simple.expense.tracker.web;

import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.springframework.web.context.request.RequestContextHolder;

import java.util.Objects;
import java.util.Optional;
import java.util.UUID;

import static org.springframework.web.context.request.RequestAttributes.SCOPE_REQUEST;

@Slf4j
public record CurrentRequestContext(
    UUID requestId
) {

    public static final String CURRENT_REQUEST_CONTEXT = "X-CURRENT-REQUEST-CONTEXT";

    public CurrentRequestContext {
        RequestContextHolder
            .currentRequestAttributes()
            .setAttribute(CURRENT_REQUEST_CONTEXT, this, SCOPE_REQUEST);
    }

    public static Optional<CurrentRequestContext> get() {
        try {
            val attributes = RequestContextHolder.currentRequestAttributes();
            val context = (CurrentRequestContext) attributes.getAttribute(CURRENT_REQUEST_CONTEXT, SCOPE_REQUEST);

            return Optional.ofNullable(context);
        } catch (IllegalStateException e) {
            log.trace("Cannot get CurrentRequestContext instance from the request attributes", e);
            return Optional.empty();
        }
    }

    public static String getRequestId() {
        return get()
            .map(CurrentRequestContext::requestId)
            .map(Objects::toString)
            .orElse("UNKNOWN");
    }
}

File: apps/server/src/main/kotlin/com/github/arhor/simple/expense/tracker/api/CurrentRequestContext.kt
package com.github.arhor.simple.expense.tracker.api

import org.slf4j.LoggerFactory
import org.springframework.web.context.request.RequestAttributes.SCOPE_REQUEST
import org.springframework.web.context.request.RequestContextHolder
import java.lang.invoke.MethodHandles
import java.util.UUID

data class CurrentRequestContext(
    val requestId: UUID,
) {
    init {
        RequestContextHolder
            .currentRequestAttributes()
            .setAttribute(CURRENT_REQUEST_CONTEXT, this, SCOPE_REQUEST)
    }

    companion object {
        const val CURRENT_REQUEST_CONTEXT = "X-CURRENT-REQUEST-CONTEXT"

        private val log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass())

        fun get() = try {
            val attributes = RequestContextHolder.currentRequestAttributes()
            val context = attributes.getAttribute(CURRENT_REQUEST_CONTEXT, SCOPE_REQUEST) as CurrentRequestContext

            context
        } catch (e: IllegalStateException) {
            log.trace("Cannot get CurrentRequestContext instance from the request attributes", e)
            null
        }

        fun getRequestId(): String {
            return get()?.requestId?.toString()
                ?: "UNKNOWN"
        }
    }
}
